ambry-router/src/main/java/com.github.ambry.router/PutManager.java

```diff
@@ -157,7 +162,12 @@ void handleResponse(ResponseInfo responseInfo) {
     PutOperation putOperation = correlationIdToPutOperation.remove(correlationId);
     // If it is still an active operation, hand over the response. Otherwise, ignore.
     if (putOperations.contains(putOperation)) {
-      putOperation.handleResponse(responseInfo);
+      try {
+        putOperation.handleResponse(responseInfo);
+      } catch (Exception e) {
+        putOperation.setOperationException(new RouterException("Put handleResponse encountered unexpected error", e,
+            RouterErrorCode.UnexpectedInternalError));
+      }
       if (putOperation.isOperationComplete() && putOperations.remove(putOperation)) {
         onComplete(putOperation);
```

Comment:
you may also have to handle exceptions coming in from `onComplete()` because of the fact that the caller callback is invoked from there?